### PR TITLE
feat(enhancement): Visual Shield Effects

### PR DESCRIPTION
**Pull request by Koranir**
## Feature Details
This feature adds a customisable, on-hit shield effect .

TODO:
- [x] Make the effect toggleable
- [x] Colors based off generators
- [x] Default to government colours if none else
- [x] Customisable shield styles
- [x] Add custom colours for all races
- [x] Add those colours to all ships
- [x] Optimise

~(third version lmao)~

## UI Screenshots

Screenshot from Screencast from 2023-07-29 17-29-03 webm

Fancy/Fast options:

https://user-images.githubusercontent.com/101683475/219926408-1143f784-0fde-4bf4-b816-3adc34e8e461.mp4

https://user-images.githubusercontent.com/101683475/225883616-45a936dd-ff78-465b-9c4e-ddb9b179739b.mp4

Custom shield textures:

https://user-images.githubusercontent.com/101683475/226147562-5db1bac8-5a81-435e-90a2-4febfc51e63f.mp4

## Usage Examples

Define color (for now in a new `shield types.txt`)
```
color "shields default" .43 .55 .70 .75
color "shields heliarch" .69 .67 .23 .75
``` 

Then in the ship or outfit definition add the name and the weight.
```
ship "Marauder Arrow"
	sprite "ship/marrows"
	thumbnail "thumbnail/marrows"
	attributes
		category "Interceptor"
		"cost" 1320000
		"shields" 2300
		"hull" 500
                "shield color" "shields default" 10
```
```
outfit "Hai Diamond Regenerator"
	category "Systems"
	cost 1150000
	thumbnail "outfit/large regenerator hai"
	"mass" 49
	"outfit space" -49
	"shield generation" 2.5
	"shield energy" 2.5
	"shield color" "shields hai strong" 15
	"shield heat" 9
	"energy consumption" .9
	description "The Hai Diamond Shield Regenerator is an advanced type of shield recharging technology. Unfortunately, its power needs are quite high, and the shifts it makes to your shield matrix causes intense heat at the regenerator when your shields are charging. The Hai usually employ Williwaw Cooling when they use this regenerator."
```

change shield texture
```
ship "Marauder Arrow"
	sprite "ship/marrows"
	thumbnail "thumbnail/marrows"
        shield "effect/shield/grid"
```

That's it!

## Testing Done
A lot more needed

## Performance Impact
Quite high, which i will attempt to lower later.
can be disabled for basically zero performance impact
or can be set to fast which will vastly lower the quality of the effect but increase performance.

### CODE REVIEW

**Comment by RisingLeaf** [comment]
👀👀👀👀👀👀 I WANT THIS

seriously looks great!

File: source/ShipEffectsShader.cpp
Comment (RisingLeaf) [inline source/ShipEffectsShader.cpp:222]:
```suggestion
}



void ShipFXShader::SetCenter(Point newCenter)
{
	center = newCenter;
}



ShipFXShader::EffectItem ShipFXShader::Prepare(const Body* body, const Point& position, const vector<pair<Point, double>>* recentHits, const float zoom, const float frame, const string &shieldColor)
```

**Comment by RisingLeaf** [comment]
I think the fadeoff could be a bit quicker

File: source/Outfit.cpp
Comment (Koranir) [reply source/Outfit.cpp:295]: The shield color is from the ship, not the generator so there should not be more than one color.

Comment (RisingLeaf) [reply source/Outfit.cpp:295]: hmmm.... in that case it should not be defined in outfit, but rather in ship

File: source/ShipEffectsShader.h
Comment (Koranir) [reply source/ShipEffectsShader.h:38]: It's what the SpriteShader does, so it's consistent

File: source/ShipEffectsShader.cpp
Comment (Koranir) [reply source/ShipEffectsShader.cpp:2]: Eh, I copied basically the entirety of SpriteShader so i think it's reasonable

File: source/ShipEffectsShader.h
Comment (RisingLeaf) [reply source/ShipEffectsShader.h:38]: hrm, ok I guess

File: source/ShipEffectsShader.cpp
Comment (RisingLeaf) [reply source/ShipEffectsShader.cpp:2]: ok

File: source/ShipEffectsShader.h
Comment (Koranir) [reply source/ShipEffectsShader.h]: Yep

File: source/ShipEffectsShader.h
Comment (Koranir) [reply source/ShipEffectsShader.h]: Done

File: source/Engine.cpp
Comment (Koranir) [reply source/Engine.cpp]: Done

**Comment by RisingLeaf** [comment]
You need to update the EndlessSkyLib.cbp file

**Comment by Ferociousfeind** [comment]
I mentioned this on discord, shield colors could be implemented in a certain way to allow outfits to adjust a ship's colors based on attributes.

You would do this with four (or five) total attributes, a "red", a "green", a "blue", a "weight", and an optional "alpha" if alpha looks OK being adjusted (maybe "thicker" shields have higher alpha?)

The color being used now would be calculated as `"red"/"weight", "green"/"weight", "blue"/"weight", 0.75` (or `"alpha"/"weight"` as you'd expect)

This would act somewhat like the "default" alpha blending you're normally accustomed to, although you'd have to be sure to multiply your desired color amounts by your decided "weight" (otherwise it'll come out very dark).

The colors defined as examples above are `"shields default" .43 .55 .70 .75` and `"shields heliarch" .69 .67 .23 .75`. If you combined a heliarch shield outfit and a human shield outfit, ones that looked something like this...
```
outfit "heliarch shield coloration"
  cost 1000
  "shield red" 6.9
  "shield green" 6.7
  "shield blue" 2.3
  "shield weight" 10

outfit "human shield coloration"
  cost 1000
  "shield red" 4.3
  "shield green" 5.5
  "shield blue" 7.
  "shield weight" 10
```
Then what you'd get is a ship exhibiting a color as if `color .56 .61 .445 .75` (which is like if a `[.43, .55, .75, .5]` pixel was alpha-blended with a `[.69, .67, .25, .5]` pixel)

My intended use case for this whole situation would be shield capacitors mostly (different alien shields generate different colored shields, tacking on an alien's capacitor onto a human ship would create a not-human, not-that-alien color shield) but also shield generators if those are still a no-go ~~for some reason~~, with ships having a strong but not _very_ strong weight so that capacitors or generators could overpower a ship's innate color mostly, but not immediately. Something like 1000 shields having 25 weight (or less), and a 100-shield capacitor having 5 weight (or maybe more, and instead of 2.5 weight)

Sorry that this is long-winded...

**Comment by RisingLeaf** [comment]
I dont think we need five attributes, just do this: 
```
"shield color" r g b w
```

**Comment by Ferociousfeind** [comment]
That's true, you could do that, but that'd require getting into the outfitter's inner workings and changing it so that it'll properly add those together to process for shield pings later

**Comment by RisingLeaf** [comment]
> That's true, you could do that, but that'd require getting into the outfitter's inner workings and changing it so that it'll properly add those together to process for shield pings later

To be honest I think it would be sloppy to make it separate attributes

**Comment by samrocketman** [comment]
- Support an additional government-based attribute (the whole government has the shield color so that you do not need to modify specific ships).
- Custom colors could be added but not associated with a specific faction.  I would say base the colors on the faction swizzle as their default.

From discussion on discord.

**Comment by Quantumshark** [comment]
> Ideas (not requirements)
> 
>     * Support an additional government-based attribute (the whole government has the shield color so that you do not need to modify specific ships).
> 
>     * Custom colors could be added but not associated with a specific faction.  I would say base the colors on the faction swizzle as their default.
> 
> 
> From discussion on discord.

If the colours were government-based, player ships would all magically change to the same colour (whatever the escort government used) regardless of what ships they were or what outfits they were using, so I definitely think colour should be tied to either ships or outfits (of course, if its an attribute then we can do either or even both) rather than to governments.

The question I'd ask is this: should the shield colour be tied to the ship, or to its shield generators? Tying it to the ship might make more sense lore-wise, but at the same time I really like the idea of having the shields of human ships with the Syndicate regenerators actually appear visibly different to those of human ships with the standard human shield generators.

Tying colour to the outfits rather than to the ships would also fix the issue of backwards compatibility with ships in old saves, so that's another bonus.

**Comment by samrocketman** [comment]
> If the colours were government-based, player ships would all magically change to the same colour

I mentioned government-based as an addition.  It doesn't mean every government needs a reflective shield.  It also might not make sense to have to define it on every ship.  So having government as an option is useful for someone who wants their whole faction to be uniform.

**Comment by samrocketman** [comment]
> Tying colour to the outfits rather than to the ships would also avoid the issue of backwards compatibility with ships in old saves, so that's another bonus.

For older saves a data migration could occur to fill in attributes when the game loads.

File: source/Engine.cpp
Comment (RisingLeaf) [inline source/Engine.cpp:1574]: move this before the if clause so that it shows up for the player too

File: source/ShipEffectsShader.cpp
Comment (samrocketman) [inline source/ShipEffectsShader.cpp:274]:
```suggestion
	item.size = 80.;
```

While play testing I found this has undesirable visual effects on smaller ships but still kept nice visual effects on larger ships.  Setting this value to static made all ships look nice.

Maybe this should just be hard coded in the shader.

**Review by samrocketman — CHANGES_REQUESTED** [review]
While play testing this pull request.

Feature feedback
----------------

- It was a challenge to assign colors to different factions because several (human) factions reuse the same ships as merchants.  By only supporting a ship attribute for shield color it is not feasible to have different colors by human faction.
  - Suggestion: Earlier feedback I gave around supporting color by government would resolve this.
- The visible shield area is too broad.  Ideally, the front edge would be much more compact when a ship is damaged to show more emphasis that the visible shield is dynamic.  On smaller ships, this is a problem because the ship is 100% covered with the damage overlay of the visible shield.  A more compact ripple would be more dynamically pronounced on smaller ships.  I think it would alse be a bit more interesting to see the damage variance on larger ships.
  - Suggestion: See my code suggestion hard coding item.size to `80.`
- Figuring out how to change the effects was a big challenge for me.  Mainly because I'm not familiar with shaders and not familiar with a lot of the variables you used.
  - Suggestion: Please add comments/documentation to all of your variables as well as what happens when you change said variable to assist would-be contributors to your future code.

Other feedback
--------------

- Debug messages are printed in ship messages area in the lower left. These should go away before merge. Though, I understand why you're printing.  The debug messages drown out all other NPC pilot messages.
- For now, there's no visual shields in vanilla.  It may be nice to default the preference of visual shields to off.  And let the player toggle it on if they want ship effects.
  - Suggestion: If there's no setting for visual shields make the default off.
  - However, I am also very OK with this preference defaulting to On.  I like fancy setting the best.


Bugs discovered
---------------

- When ships are disabled, they sometimes keep the visual effects of shield damage always visible.  I think the shield effect should disappear because shields are down when a ship is disabled.
- When player ship is flying over another ship with its visible shield up the visible shield continues to be drawn as an outline over the player ship.  The ship being shot is underneath the player flagship.  However, the outline shows over.
- The shield effect only applies to NPCs.  The player flagship has no visible shield effect.

**Comment by RisingLeaf** [comment]
> For now, there's no visual shields in vanilla. It may be nice to default the preference of visual shields to off. And let the player toggle it on if they want ship effects.
> 
>     * Suggestion: If there's no setting for visual shields make the default off.

I would rather make the default setting on.
But that is just a personal opinion.
If you make the default off, players are less likely to discover them.

**Comment by samrocketman** [comment]
> I would rather make the default setting on.
> But that is just a personal opinion.
> If you make the default off, players are less likely to discover them

Fair I am OK with it defaulting to on as well.  I should probably better indicate my non-strong opinion about it

File: source/ShipEffectsShader.cpp
Comment (samrocketman) [inline source/ShipEffectsShader.cpp:164]:
```suggestion
		"      color += shieldColor * recentDamage[i] * clamp(2. - distance(hitPoint, uv)*.04*80, 0., 1.5);\n"
```

Instead of `item.size = 80` (from my earlier feedback) you could update the shader to hard-code it

Comment (samrocketman) [reply source/ShipEffectsShader.cpp:164]: not here

File: source/ShipEffectsShader.cpp
Comment (samrocketman) [inline source/ShipEffectsShader.cpp:110]:
```suggestion
		"uniform float size = 80;\n"
```

Instead of `item.size = 80` (from my earlier feedback) you could update the shader to hard-code it

Comment (RisingLeaf) [reply source/ShipEffectsShader.cpp:110]: +1

**Comment by Ferociousfeind** [comment]
I am going to do some testing on this PR

It'll involve making sure it all works
then discovering what different ships are colored
then testing weights and colors to see how it all interacts
then testing a few custom outfits to see what weight ratios look the way I want them to look (nominally, like, 10% extra shields from outfits getting 50% of the ship's shields' weight on color) (don't know about shield generators though, maybe like 60 seconds of shield generation = capacitor's worth of color weight)

and then I will deliver proposals for shield generators, colors and weights on them! And whether or not I think this is a great PR

**Comment by samrocketman** [comment]
I opened <https://github.com/Koranir/endless-sky/pull/14> to fix issues with this pull request.

Upstream update, preference ordering, and code style fix.

**Comment by samrocketman** [comment]
> Does it make sense for colors to be government based at all? The color of the shields is going to be a physical property of the emitters/generators, so to me it would make sense for it to solely be based on the ship and outfit attributes.

I believe they are.  If you mix outfits the colors mix.

**Comment by samrocketman** [comment]
> I see that (which makes sense), but is the information in the PR about defaulting to the government color outdated?

I defer to the author it might be an option (because I asked for it originally when I was having trouble testing if you look at my first review).

Later other people had better feedback and guidance rather than my simple gov't feedback.  I believe the latest iteration is the color mixing based on outfit.

I'm not looking at the code at the moment to be clear (on mobile).

**Comment by samrocketman** [comment]
Looking at that data files it looks like colors mix between:
- Government
- Ship
- Outfit

So there's lots of options.

**Comment by Koranir** [comment]
> Does it make sense for colors to be government based at all? The color of the shields is going to be a physical property of the emitters/generators, so to me it would make sense for it to solely be based on the ship and outfit attributes.

The government color is the fallback, weighted very lightly so that any ship or outfit shield colour basically overrides 90% of it.

**Comment by bene-dictator** [comment]
> Does it make sense for colors to be government based at all? The color of the shields is going to be a physical property of the emitters/generators, so to me it would make sense for it to solely be based on the ship and outfit attributes.

The ship has its own shield emitters that are replenished every time you land, but in space they're replenished by your generator. The emitters are independent of the generator and are installed by the ship's designers. So it makes sense that the fallback is the government that created the ship.

**Comment by 1010todd** [comment]
I played around with this via Sam's fork and it isn't really that noticable (at least on fancy setting). I noticed you have new image and popped that in but I still feel like it need more contrast.

Also is it possible to change the shield texture per generator/government? I think this is definitely something people would want to change for their race/shield gen.

**Comment by 1010todd** [comment]
Shield seems to be offset from the edge on my ship for some reason. Also it seems to not support more than one impact. (Also I adjusted the shield texture a bit.)

image

**Comment by samrocketman** [comment]
@1010todd worth noting my fork doesn't have the last two commits, yet (contrast commits).  I can let you know when I update (likely today)

**Comment by Koranir** [comment]
> I played around with this via Sam's fork and it isn't really that noticable (at least on fancy setting). I noticed you have new image and popped that in but I still feel like it need more contrast.
> 
> Also is it possible to change the shield texture per generator/government? I think this is definitely something people would want to change for their race/shield gen.

Pretty sure shilds are very tied to the ships they are on, so the shield texture is changeable for individual ships.

**Comment by samrocketman** [comment]
https://github.com/Koranir/endless-sky/pull/15 style fix

**Review by pilover100 — COMMENTED** [review]
This is definately a much better and neater solution to my PR for ships being hit and I very much like the concept this PR puts forwards. However, I do have a question: I know that shield colours are most heavilly weighted towards installed outfits and attributes on the ships themselves to which I agree (they should indeed contribute to the shield colour). But what was the rational towards having the republic use a slightly different shield colour purely because the ship just happens to currently be opperated by the republic?

image

When I look at the colours for the republic (left) and standard (right) shields, there isn't much of a major distinction to my eye, and is not likely going to be noticed much especially in the middle of a fight in which republic ships are going to be targeted/hit (FW, the alpha attack on farpoint, fighting pirates).

Additionally, if the player captures a navy ship (which many may do in the FW campaign) having the shield colour suddenly shift dispite no change to the actual tech used with the vessel makes very little sense to me.

Other than that, this is a good step forwards in how ships deal with receving damage and convaying it to the player, and a much better solution than the one I put forwards in my PR.

**Comment by Koranir** [comment]
> This is definately a much better and neater solution to my PR for ships being hit and I very much like the concept this PR puts forwards. However, I do have a question: I know that shield colours are most heavilly weighted towards installed outfits and attributes on the ships themselves to which I agree (they should indeed contribute to the shield colour). But what was the rational towards having the republic use a slightly different shield colour purely because the ship just happens to currently be opperated by the republic?


It just feels right

**Comment by RisingLeaf** [comment]
One of the grid textures is way too high resolution it just needs to be twice the size of its display on screen

**Comment by Quantumshark** [comment]
The Navy have their own suite of ships, and rarely if ever use any not from their set. So you could put the republic colour on Navy ships, hinting that maybe they use more advanced shield technology than is available elsewhere in human space, rather than on the government.

**Comment by Amazinite** [comment]
Do the two TODO items in the PR description still need to be completed?

**Comment by Koranir** [comment]
All cleaned up and ready for review.

File: source/Ship.h
Comment (RisingLeaf) [inline source/Ship.h:373]: So for hazards it is system center?
If yes, that would probably contradict hazards with "constant strength"
But that's only a minor detail

File: source/ShipEffectsShader.h
Comment (RisingLeaf) [inline source/ShipEffectsShader.h:30]: Two too many linebreaks

File: source/ShipEffectsShader.cpp
Comment (RisingLeaf) [inline source/ShipEffectsShader.cpp:91]: Reason for the rename?

File: source/ShipEffectsShader.cpp
Comment (RisingLeaf) [inline source/ShipEffectsShader.cpp:182]: The shader code would profit from some comments

**Review by RisingLeaf — COMMENTED** [review]
Please don't expect replies to replies for the next couple weeks. That's why I am just commenting.

One thing for the fast option to consider, render sprites first then effects so that you don't have to bind/unbind the whole time.

**Comment by bene-dictator** [comment]
Any updates on this? Would love to see it in main : )

**Comment by Koranir** [comment]
> Any updates on this? Would love to see it in main : )

I'm in the midst of rewriting this. Might take a while, I'm in the middle of exams right now.

**Comment by Saugia** [comment]
Still plan on coming back to this soon? If not, we could just close it for the time being until it is caught up. Definitely something we're looking at though.

**Comment by Koranir** [comment]
Will open a new PR one day with the PR rewritten.

**Comment by samrocketman** [comment]
This would be awesome :-)

**Pull request closed without merging**
